modules/460-log-shipper/hooks/internal/handler/generator.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2021 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/v1alpha1"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/vector/config"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/vector/model"
+)
+
+type Generator struct {
+	logDestinations map[string]v1alpha1.ClusterLogDestination
+}
+
+func FromInput(input *go_hook.HookInput) *Generator {
+	destNameToObject := make(map[string]v1alpha1.ClusterLogDestination)
+	snap := input.Snapshots["cluster_log_destination"]
+
+	// load all destinations into the destination map
+	for _, d := range snap {
+		dest := d.(v1alpha1.ClusterLogDestination)
+		destNameToObject[dest.Name] = dest
+	}
+
+	return &Generator{logDestinations: destNameToObject}
+}
+
+func (g *Generator) Do(input *go_hook.HookInput) ([]byte, error) {
```

Comment:
Yeah, I thought about something like this. The main reason is that in my mind, `FomInput` is one way to create a new Generator instance. Theoretically, there can be others like `FromYAML` or `FromMap`.

It looks like this because I try to keep as much logic as possible from the original hook. Then we can write more granular unit tests and change the structure.